weld/syntax/macro_processor.rs

```diff
@@ -136,67 +212,141 @@ fn update_defined_ids(expr: &mut Expr, sym_gen: &mut SymbolGenerator) {
     assert_eq!(print_expr_without_indent(&result).as_str(), "((b*b)+(b*b))");
 
     let macros = parse_macros("macro foo(a) = a + a;").unwrap();
     let expr = parse_expr("foo(foo(b))").unwrap();
-    let result = process_expression(&expr, &macros).unwrap();
+    let result = process_macros(&expr, &macros).unwrap();
     assert_eq!(print_expr_without_indent(&result).as_str(), "((b+b)+(b+b))");
 
     // Macro whose parameter symbol is also used as an argument inside
     let macros = parse_macros("macro foo(a) = a + a;").unwrap();
     let expr = parse_expr("foo(a * a)").unwrap();
-    let result = process_expression(&expr, &macros).unwrap();
+    let result = process_macros(&expr, &macros).unwrap();
     assert_eq!(print_expr_without_indent(&result).as_str(), "((a*a)+(a*a))");
 
     // Multiple macros
     let macros = parse_macros("macro foo(a) = a + a; macro bar(a, b) = a * b;").unwrap();
     let expr = parse_expr("foo(bar(a, b))").unwrap();
-    let result = process_expression(&expr, &macros).unwrap();
+    let result = process_macros(&expr, &macros).unwrap();
     assert_eq!(print_expr_without_indent(&result).as_str(), "((a*b)+(a*b))");
 }
 
+#[test]
+fn basic_type_alias() {
+    let aliases = parse_type_aliases("type int = i32;").unwrap();
+    let mut expr = parse_expr("|a: int| a + 1").unwrap();
+    process_type_aliases(&mut expr, aliases).unwrap();
+    assert_eq!(print_expr_without_indent(&expr).as_str(), "|a:i32|(a+1)");
+}
+
+#[test]
+fn overwrite_type_alias() {
+    let aliases = parse_type_aliases("type int = i32; type int = i64;").unwrap();
+    let mut expr = parse_expr("|a: int| a + 1L").unwrap();
+    process_type_aliases(&mut expr, aliases).unwrap();
+    assert_eq!(print_expr_without_indent(&expr).as_str(), "|a:i64|(a+1L)");
+}
+
+#[test]
+fn invalid_type_alias() {
+    let aliases = parse_type_aliases("type int = i32;").unwrap();
+    let mut expr = parse_expr("|a: otherInt| a + 1").unwrap();
+    process_type_aliases(&mut expr, aliases).expect_err(
+        "Type alises processing should fail with unknown type"
+    );
+}
+
+#[test]
+fn type_alias_in_type_alias() {
+    let aliases = parse_type_aliases("type int = i32; type pair = {int,int};").unwrap();
+    let mut expr = parse_expr("|a: pair| a.$0 + 1").unwrap();
+    process_type_aliases(&mut expr, aliases).unwrap();
+    assert_eq!(print_expr_without_indent(&expr).as_str(), "|a:{i32,i32}|(a.$0+1)");
+}
+
+#[test]
+fn redefinition_with_type_alias_in_type_alias() {
+    // Types are defined in order: redefining a type shouldn't redefine
+    // types that occur before it.
+    let aliases = parse_type_aliases("type int = i32;
+                                     type pair = {int,int};
+                                     type int = i64;").unwrap();
+    let mut expr = parse_expr("|a: pair| a").unwrap();
+    process_type_aliases(&mut expr, aliases).unwrap();
+    assert_eq!(print_expr_without_indent(&expr).as_str(), "|a:{i32,i32}|a");
```

Comment:
suggest adding:
```
  let mut expr = parse_expr("|b: int| b").unwrap();
  process_type_aliases(&mut expr, aliases).unwrap();
  assert_eq!(print_expr_without_indent(&expr).as_str(), "|b:i64|b");
```
if that is indeed the intended behavior

Reply:
I think it makes more sense for the definitions to be processed in order (there's actually some extra logic I added to make sure that was the case) -- otherwise it becomes confusing since later definitions modify earlier ones.